validator/sawtooth_validator/journal/chain_commit_state.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2016 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+
+class _CommitCache(object):
+    """Tracks the commit status of a set of identifiers and these identifiers
+    are either explicitly committed, or explicitly uncommitted. If they fall in
+    to neither of these cases then the fallback is to look in the BlockStore to
+    see if they are there. Explicit committed ids take priority over
+    uncommitted since one of the common use cases we have is to simulate the
+    committed state at a previous state of the BlockStore and we allow for the
+    identifiers to be re-committed.
+    """
+    def __init__(self, block_store_check):
+        self.block_store_check = block_store_check
+        self._committed = set()  # the set of items
+        # committed by this chain
+        self._uncommitted = set()  # the set of items
+        # uncommitted by the current chain when it is rolled back.
+
+    def add(self, identifier):
+        self._committed.add(identifier)
+
+    def remove(self, identifier):
+        self._committed.discard(identifier)
+
+    def uncommit(self, identifier):
+        self._uncommitted.add(identifier)
+
+    def __contains__(self, identifier):
+        if identifier in self._committed:
+            return True
+        elif identifier in self._uncommitted:
+            return False
+        return self.block_store_check(identifier)
+
+
+class ChainCommitState(object):
```

Comment:
The BatchTrack and the ChainCommitState provide different functionality. The BatchTracker appears to be attepting to provide notifications when a batch appears in the BlockStore. The ChainCommitState is focused on simulating an alternate version of the the BlockStore.